janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java

```diff
@@ -0,0 +1,460 @@
+    }
+
+    @Override
+    public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) throws BackendException {
+        if (this.storeManager.getFeatures().hasOrderedScan()) {
+            throw new PermanentBackendException("This operation is only allowed when a random partitioner (md5 or murmur3) is used.");
+        }
+
+        return Try.of(() -> new ResultSetKeyIterator(
+                query,
+                this.getter,
+                this.session.execute(this.getKeysAll.bind()
+                        .setBytes(SLICE_START_BINDING, query.getSliceStart().asByteBuffer())
+                        .setBytes(SLICE_END_BINDING, query.getSliceEnd().asByteBuffer())
+                        .setFetchSize(this.storeManager.getPageSize())
+                        .setConsistencyLevel(getTransaction(txh).getReadConsistencyLevel()))))
+                .getOrElseThrow(EXCEPTION_MAPPER);
+    }
+
+    // ------------------------------------------------------------------------
+    // Inner classes
+    // ------------------------------------------------------------------------
+
+    static class ResultSetKeyIterator extends AbstractIterator<StaticBuffer> implements KeyIterator {
+
+        private final SliceQuery sliceQuery;
+        private final Getter getter;
+        private final Iterator<Row> iterator;
+
+        private Row currentRow = null;
+        private StaticBuffer currentKey = null;
+        private StaticBuffer lastKey = null;
+
+        ResultSetKeyIterator(final SliceQuery sliceQuery, final Getter getter, final ResultSet resultSet) {
+            this.sliceQuery = sliceQuery;
+            this.getter = getter;
+            this.iterator = Iterator.ofAll(resultSet.iterator())
+                    .peek(row -> {
+                        this.currentRow = row;
+                        this.currentKey = StaticArrayBuffer.of(row.getBytes(KEY_COLUMN_NAME));
+                    });
+        }
+
+        @Override
+        protected StaticBuffer computeNext() {
+            if (this.currentKey != null && !this.currentKey.equals(this.lastKey)) {
+                this.lastKey = this.currentKey;
+                return this.lastKey;
+            }
+
+            while (this.iterator.hasNext()) {
+                this.iterator.next();
+                if (!this.currentKey.equals(this.lastKey)) {
+                    this.lastKey = this.currentKey;
+                    return this.lastKey;
+                }
+            }
+            return endOfData();
+        }
+
+        @Override
+        public RecordIterator<Entry> getEntries() {
+            return new EntryRecordIterator(this.sliceQuery, this.getter, Iterator.of(this.currentRow).concat(this.iterator), this.currentKey);
+        }
+
+        @Override
+        public void close() throws IOException {
+            // NOP
+        }
+
+        static class EntryRecordIterator extends AbstractIterator<Entry> implements RecordIterator<Entry> {
```

Comment:
I'd like to leave this one as an inner class of the Itertor and make it private as it's really an internal implementation detail.